README.md
# lightning-sys

[![crates.io](https://img.shields.io/crates/v/lightning-sys.svg)](https://crates.io/crates/lightning-sys)
[![docs.rs](https://docs.rs/lightning-sys/badge.svg)](https://docs.rs/lightning-sys/)
[![Build Status](https://travis-ci.com/Petelliott/lightning-sys.svg?branch=master)](https://travis-ci.com/Petelliott/lightning-sys)

Safe gnu lightning bindings for rust

## GNU lightning version

This crate links explicitly against its own packaged copy of lightning-2.1.3.

## MSRV

The minimum supported Rust version is **1.40**.

## Examples

### a function that increments a number by one
```rust
use lightning_sys::{Jit, Reg, JitPointer, JitWord};

let mut jit = Jit::new();
let mut js = jit.new_state();

js.prolog();
let inarg = js.arg();
js.getarg(Reg::R(0), &inarg);
js.addi(Reg::R(0), Reg::R(0), 1);
js.retr(Reg::R(0));

let incr = unsafe { js.cast_emit::<extern fn(JitWord) -> JitWord>() };
js.clear_state();

assert_eq!(incr(5), 6);
assert_eq!(incr(6), 7);

```

### A simple function call to `printf`
```rust
extern crate libc;

use std::ffi::CString;
use lightning_sys::{Jit, JitWord, Reg, JitPointer};
use std::convert::TryInto;

fn main() {
    let mut jit = Jit::new();
    let mut js = jit.new_state();

    // make sure this outlives any calls
    let cs = CString::new("generated %d bytes\n").unwrap();

    let start = js.note(file!(), line!());
    js.prolog();
    let inarg = js.arg();
    js.getarg(Reg::R(1), &inarg);
    js.prepare();
    js.pushargi(cs.as_ptr() as JitWord);
    js.ellipsis();
    js.pushargr(Reg::R(1));
    js.finishi(libc::printf as JitPointer);
    js.ret();
    js.epilog();
    let end = js.note(file!(), line!());

    let my_function = unsafe{ js.cast_emit::<extern fn(JitWord)>() };
    /* call the generated code, passing its size as argument */
    my_function((js.address(&end) as u64 - js.address(&start) as u64).try_into().unwrap());
    js.clear_state();

    // TODO: dissasembly has not been implemented yet
    // js.dissasemble();
}

```
### Fibonacci numbers
```rust
use lightning_sys::{Jit, JitWord, Reg, JitPointer, NULL};

fn main() {
    let mut jit = Jit::new();
    let mut js = jit.new_state();

    let label = js.label();
                js.prolog();
    let inarg = js.arg();
                js.getarg(Reg::R(0), &inarg);
    let zero  = js.beqi(Reg::R(0), 0);
                js.movr(Reg::V(0), Reg::R(0));
                js.movi(Reg::R(0), 1);
    let refr  = js.blei(Reg::V(0), 2);
                js.subi(Reg::V(1), Reg::V(0), 1);
                js.subi(Reg::V(2), Reg::V(0), 2);
                js.prepare();
                js.pushargr(Reg::V(1));
    let call  = js.finishi(NULL);
                js.patch_at(&call, &label);
                js.retval(Reg::V(1));
                js.prepare();
                js.pushargr(Reg::V(2));
    let call2 = js.finishi(NULL);
                js.patch_at(&call2, &label);
                js.retval(Reg::R(0));
                js.addr(Reg::R(0), Reg::R(0), Reg::V(1));

                js.patch(&refr);
                js.patch(&zero);
                js.retr(Reg::R(0));
                js.epilog();

    let fib = unsafe{ js.cast_emit::<extern fn(JitWord) -> JitWord>() };
    js.clear_state();

    println!("fib({})={}", 32, fib(32));
    assert_eq!(0, fib(0));
    assert_eq!(1, fib(1));
    assert_eq!(1, fib(2));
    assert_eq!(2178309, fib(32));
}
```

### Tail Call Optimized factorial
```rust
use lightning_sys::{Jit, JitWord, Reg, NULL};

fn main() {
    let mut jit = Jit::new();
    let mut js = jit.new_state();

    let fact = js.forward();

                js.prolog();
    let inarg = js.arg();
                js.getarg(Reg::R(0), &inarg);
                js.prepare();
                js.pushargi(1);
                js.pushargr(Reg::R(0));
    let call  = js.finishi(NULL);
                js.patch_at(&call, &fact);

                js.retval(Reg::R(0));
                js.retr(Reg::R(0));
                js.epilog();

    js.link(&fact);
                js.prolog();
                js.frame(16);
    let f_ent = js.label(); // TCO entry point
    let ac    = js.arg();
    let ina   = js.arg();
                js.getarg(Reg::R(0), &ac);
                js.getarg(Reg::R(1), &ina);
    let f_out = js.blei(Reg::R(1), 1);
                js.mulr(Reg::R(0), Reg::R(0), Reg::R(1));
                js.putargr(Reg::R(0), &ac);
                js.subi(Reg::R(1), Reg::R(1), 1);
                js.putargr(Reg::R(1), &ina);
    let jump  = js.jmpi(); // tail call optimiation
                js.patch_at(&jump, &f_ent);
                js.patch(&f_out);
                js.retr(Reg::R(0));

    let factorial = unsafe{ js.cast_emit::<extern fn(JitWord) -> JitWord>() };
    js.clear_state();

    println!("factorial({}) = {}", 5, factorial(5));
}
```


Changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Changed
- Set a Minimum Supported Rust Version (MSRV) of 1.40.0 (#49)

## [0.2.2] - 2020-08-22
### Added
- Introduced `Changelog.md` (#48)
- Set up CI builds for multiple major architectures and OSes (#15, #18, #27)
- Introduced basic code coverage metrics with Tarpaulin and Codecov (#24)
- Ported remaining examples from GNU lightning (#28, #31)
- Laid groundwork for interface generation mechanization (#32)
- Added some missing entry points (#34, #35, #37, #42)

### Changed
- Allow multiple `JitState`s to exist at once (#14)
- Use GNU Lightning 2.1.3 (#19)
- Switched back to static linking for liblightning (#23)
- Set a Minimum Supported Rust Version (MSRV) of 1.39.0 (#41)
- Switched back to a Git submodule for GNU lightning (#45)
- Made various build and code quality improvements (#10, #11, #17, #20, #25, #26, #36, #40, #46)

### Fixed
- Corrected a register reference in Fibonacci example (#16)
- Corrected lifetimes and made API semantically mutating (#30)

## [0.2.1] - 2020-05-08
### Added
- Added assertions to factorial example

### Changed
- Corrected link to GitHub repo

## [0.2.0] - 2019-08-17
### Added
- Added support for f32 and f64 entry points
- Added predicate functions `forward_p`, `indirect_p`, `target_p`
- Added the tail-call-optimized factorial example from GNU lightning

### Changed
- Relicensed under LGPL (from GPL), matching GNU lightning

## [0.1.2] - 2019-08-13
### Added
- Added Fibonacci example from GNU lightning
- Added branch/jump instructions
- Added (`panic`-ing) bounds checks to guard against invalid register indices
- Introduced aliases for entry points that redirect to other entry points

### Fixed
- Fixed some bugs uncovered by Fibonacci example

## [0.1.1] - 2019-08-12
Early preview release

## [0.1.0] - 2019-07-21
Initial (non-functional) release

[Unreleased]: https://github.com/petelliott/lightning-sys/compare/v0.2.2...HEAD
[0.2.2]: https://github.com/petelliott/lightning-sys/compare/v0.2.1...v0.2.2
[0.2.1]: https://github.com/petelliott/lightning-sys/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/petelliott/lightning-sys/compare/v0.1.2...v0.2.0
[0.1.2]: https://github.com/petelliott/lightning-sys/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/petelliott/lightning-sys/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/petelliott/lightning-sys/releases/tag/v0.1.0


src/lib.rs
//! lightning-sys aims provides safe (as safe as a jit can be) rust bindings to GNU Lightning
//!
//! ## Examples:
//! ### a function that increments a number by one
//! ```
//! use lightning_sys::{Jit, Reg, JitPointer, JitWord};
//!
//! let mut jit = Jit::new();
//! let mut js = jit.new_state();
//!
//! js.prolog();
//! let inarg = js.arg();
//! js.getarg(Reg::R(0), &inarg);
//! js.addi(Reg::R(0), Reg::R(0), 1);
//! js.retr(Reg::R(0));
//!
//! let incr = unsafe { js.cast_emit::<extern fn(JitWord) -> JitWord>() };
//! js.clear_state();
//!
//! assert_eq!(incr(5), 6);
//! assert_eq!(incr(6), 7);
//!
//! ```
//!
//! ### A simple function call to `printf`
//! ```
//! extern crate libc;
//!
//! use std::ffi::CString;
//! use lightning_sys::{Jit, JitWord, Reg, JitPointer};
//! use std::convert::TryInto;
//!
//! fn main() {
//!     let mut jit = Jit::new();
//!     let mut js = jit.new_state();
//!
//!     // make sure this outlives any calls
//!     let cs = CString::new("generated %d bytes\n").unwrap();
//!
//!     let start = js.note(Some(file!()), line!());
//!     js.prolog();
//!     let inarg = js.arg();
//!     js.getarg(Reg::R(1), &inarg);
//!     js.prepare();
//!     js.pushargi(cs.as_ptr() as JitWord);
//!     js.ellipsis();
//!     js.pushargr(Reg::R(1));
//!     js.finishi(libc::printf as JitPointer);
//!     js.ret();
//!     js.epilog();
//!     let end = js.note(Some(file!()), line!());
//!
//!     let my_function = unsafe{ js.cast_emit::<extern fn(JitWord)>() };
//!     /* call the generated code, passing its size as argument */
//!     my_function((js.address(&end) as u64 - js.address(&start) as u64).try_into().unwrap());
//!     js.clear_state();
//!
//!     // TODO: dissasembly has not been implemented yet
//!     // js.dissasemble();
//! }
//!
//! ```
//!
//! ### Fibonacci numbers
//! ```
//! use lightning_sys::{Jit, JitWord, Reg, JitPointer, NULL};
//!
//! fn main() {
//!     let mut jit = Jit::new();
//!     let mut js = jit.new_state();
//!
//!     let label = js.label();
//!                 js.prolog();
//!     let inarg = js.arg();
//!                 js.getarg(Reg::R(0), &inarg);
//!     let zero  = js.beqi(Reg::R(0), 0);
//!                 js.movr(Reg::V(0), Reg::R(0));
//!                 js.movi(Reg::R(0), 1);
//!     let refr  = js.blei(Reg::V(0), 2);
//!                 js.subi(Reg::V(1), Reg::V(0), 1);
//!                 js.subi(Reg::V(2), Reg::V(0), 2);
//!                 js.prepare();
//!                 js.pushargr(Reg::V(1));
//!     let call  = js.finishi(NULL);
//!                 js.patch_at(&call, &label);
//!                 js.retval(Reg::V(1));
//!                 js.prepare();
//!                 js.pushargr(Reg::V(2));
//!     let call2 = js.finishi(NULL);
//!                 js.patch_at(&call2, &label);
//!                 js.retval(Reg::R(0));
//!                 js.addr(Reg::R(0), Reg::R(0), Reg::V(1));
//!
//!                 js.patch(&refr);
//!                 js.patch(&zero);
//!                 js.retr(Reg::R(0));
//!                 js.epilog();
//!
//!     let fib = unsafe{ js.cast_emit::<extern fn(JitWord) -> JitWord>() };
//!     js.clear_state();
//!
//!     println!("fib({})={}", 32, fib(32));
//!     assert_eq!(0, fib(0));
//!     assert_eq!(1, fib(1));
//!     assert_eq!(1, fib(2));
//!     assert_eq!(2178309, fib(32));
//! }
//! ```
//!
//! ### Tail Call Optimized factorial
//! ```
//! use lightning_sys::{Jit, JitWord, Reg, NULL};
//!
//! fn main() {
//!     let mut jit = Jit::new();
//!     let mut js = jit.new_state();
//!
//!     let fact = js.forward();
//!
//!                 js.prolog();
//!     let inarg = js.arg();
//!                 js.getarg(Reg::R(0), &inarg);
//!                 js.prepare();
//!                 js.pushargi(1);
//!                 js.pushargr(Reg::R(0));
//!     let call  = js.finishi(NULL);
//!                 js.patch_at(&call, &fact);
//!
//!                 js.retval(Reg::R(0));
//!                 js.retr(Reg::R(0));
//!                 js.epilog();
//!
//!     js.link(&fact);
//!                 js.prolog();
//!                 js.frame(16);
//!     let f_ent = js.label(); // TCO entry point
//!     let ac    = js.arg();
//!     let ina   = js.arg();
//!                 js.getarg(Reg::R(0), &ac);
//!                 js.getarg(Reg::R(1), &ina);
//!     let f_out = js.blei(Reg::R(1), 1);
//!                 js.mulr(Reg::R(0), Reg::R(0), Reg::R(1));
//!                 js.putargr(Reg::R(0), &ac);
//!                 js.subi(Reg::R(1), Reg::R(1), 1);
//!                 js.putargr(Reg::R(1), &ina);
//!     let jump  = js.jmpi(); // tail call optimiation
//!                 js.patch_at(&jump, &f_ent);
//!                 js.patch(&f_out);
//!                 js.retr(Reg::R(0));
//!
//!     let factorial = unsafe{ js.cast_emit::<extern fn(JitWord) -> JitWord>() };
//!     js.clear_state();
//!
//!     println!("factorial({}) = {}", 5, factorial(5));
//!     assert_eq!(1, factorial(1));
//!     assert_eq!(2, factorial(2));
//!     assert_eq!(6, factorial(3));
//!     assert_eq!(24, factorial(4));
//!     assert_eq!(120, factorial(5));
//! }
//! ```
// Enforce some lints for the whole crate.
#![deny(clippy::needless_lifetimes)]
#![deny(clippy::transmute_ptr_to_ptr)]
#![allow(clippy::needless_doctest_main)] // remain faithful to original examples

// Suppress some lints for bindings specifically.
#[allow(non_upper_case_globals)]
#[allow(non_camel_case_types)]
#[allow(dead_code)]
#[allow(clippy::unreadable_literal)]
#[allow(clippy::trivially_copy_pass_by_ref)]
#[allow(clippy::useless_transmute)]
#[allow(clippy::too_many_arguments)]
mod bindings;

#[macro_use]
extern crate lazy_static;

// The `raw` module comes first so that it can provide macros during parsing of
// other modules. It is not yet part of the crate's public API, but in the
// spirit of *-sys modules exposing lowest-level bindings, it may someday be so.
#[macro_use]
pub(crate) mod raw;

pub mod jit;
pub use jit::Jit;

pub mod jitstate;
pub use jitstate::JitState;

pub mod types;
pub use types::NULL;
pub use types::Reg;
pub use types::JitNode;
pub use types::{JitWord, JitUword, JitPointer};
pub(crate) use types::ToFFI;
